src/Data/Singletons/Promote.hs

```diff
@@ -325,14 +328,56 @@ promoteInstanceDec meth_sigs
 
     lookup_cls_tvb_names :: PrM [Name]
     lookup_cls_tvb_names = do
-      mb_info <- dsReify pClsName
+      let mk_tvb_names = extract_tvb_names (dsReify pClsName)
+                     <|> extract_tvb_names (dsReify cls_name)
+                     <|> extract_tvb_names (dsReifyTypeNameInfo pClsName)
+                     <|> extract_tvb_names (dsReifyTypeNameInfo cls_name)
+                      -- See Note [Using dsReifyTypeNameInfo when promoting instances]
+      mb_tvb_names <- runMaybeT mk_tvb_names
+      case mb_tvb_names of
+        Just tvb_names -> pure tvb_names
+        Nothing -> fail $ "Cannot find class declaration annotation for " ++ show cls_name
+
+    extract_tvb_names :: PrM (Maybe DInfo) -> MaybeT PrM [Name]
+    extract_tvb_names reify_info = do
+      mb_info <- lift reify_info
       case mb_info of
-        Just (DTyConI (DClassD _ _ tvbs _ _) _) -> return (map extractTvbName tvbs)
-        _ -> do
-          mb_info' <- dsReify cls_name
-          case mb_info' of
-            Just (DTyConI (DClassD _ _ tvbs _ _) _) -> return (map extractTvbName tvbs)
-            _ -> fail $ "Cannot find class declaration annotation for " ++ show cls_name
+        Just (DTyConI (DClassD _ _ tvbs _ _) _)
+          -> pure $ map extractTvbName tvbs
+        _ -> empty
+
+{-
+Note [Using dsReifyTypeNameInfo when promoting instances]
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+During the promotion of a class instance, it becomes necessary to reify the
+original promoted class's info to learn various things. It's tempting to think
+that just calling dsReify on the class name will be sufficient, but it's not.
+Consider this class and its promotion:
+
+  class Eq a where
+    (==) :: a -> a -> Bool
+
+  class PEq a where
+    type (==) (x :: a) (y :: a) :: Bool
+
+Notice how both of these classes have an identifier named (==), one at the
+value level, and one at the type level. Now imagine what happens when you
+attempt to promote this Template Haskell declaration:
+
+   [d| f :: Bool
+       f = () == () |]
+
+Because it's quasiquoted, Template Haskell will attempt to reify a raw "=="
+Name. But due to GHC Trac #13961, Template Haskell has to arbitrarily choose
+between reifying the info for the value-level (==) and the type-level (==), and
+in this case, it happens to pick the value-level (==) info. We want the type-level
+(==) info, however, because we care about the promoted version of (==).
+
+Fortunately, there's a serviceable workaround. After calling dsReify, if we happen
```

Comment:
Actually, I believe I did misunderstand your "Would the bottom two always work?" question. I think I should have interpreted that as "Can you scrap the top two alternatives in favor of just the bottom two alternatives?" And the answer to that question is yes, happily enough. I've applied this change just now.